presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -484,6 +484,20 @@ public PlanWithProperties visitSort(SortNode node, Context context)
                         gatheringExchange(idAllocator.getNextId(), REMOTE, child.getNode()),
                         child.getProperties());
             }
+            else {
+                // current plan so far is single node, so local properties are effectively global properties
+                // skip the SortNode if the local properties guarantee ordering on Sort keys
+                // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator
+                List<LocalProperty<Symbol>> desiredProperties = new ArrayList<>();
+                for (Symbol symbol : node.getOrderBy()) {
+                    desiredProperties.add(new SortingProperty<>(symbol, node.getOrderings().get(symbol)));
+                }
+
+                if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
+                        .noneMatch(Optional::isPresent)) {
+                    return child;
+                }
+            }
```

Comment:
is it possible to make a separate optimizer for this?

I don't like the idea `AddExchanges` does such optimizations. IMO `AddExchanges` should only (as name suggest) add `ExchangeNode` into the plan.